src/mysql-util/src/privileges.rs

```diff
@@ -0,0 +1,235 @@
+    let mut errors = tables
+        .iter()
+        .flat_map(|table| {
+            ["LOCK TABLES", "SELECT"].iter().filter_map(|privilege| {
+                // Check both the wildcard schema and the specific schema
+                let privileged = [grant_map.get("*"), grant_map.get(table.schema_name)]
+                    .iter()
+                    .any(|schema_map| {
+                        // Check both the wildcard table and the specific table
+                        schema_map.map_or(false, |schema_map| {
+                            [schema_map.get("*"), schema_map.get(table.table_name)]
+                                .iter()
+                                .any(|privs| {
+                                    privs.map_or(false, |privs| {
+                                        privs.contains(*privilege)
+                                            || privs.contains("ALL PRIVILEGES")
+                                    })
+                                })
+                        })
+                    });
+
+                if !privileged {
+                    Some(MissingPrivilege {
+                        privilege: privilege.to_string(),
+                        qualified_table_name: format!("{}.{}", table.schema_name, table.table_name),
+                    })
+                } else {
+                    None
+                }
+            })
+        })
+        .collect::<Vec<_>>();
+
+    // Check that we have REPLICATION SLAVE priviliges at the global level
+    if !grant_map
+        .get("*")
+        .and_then(|schema_map| schema_map.get("*"))
+        .map(|privs| privs.contains("REPLICATION SLAVE") || privs.contains("ALL PRIVILEGES"))
+        .unwrap_or(false)
+    {
+        errors.push(MissingPrivilege {
+            privilege: "REPLICATION SLAVE".to_string(),
+            qualified_table_name: "*.*".to_string(),
+        });
+    }
+
+    if !errors.is_empty() {
+        Err(MySqlError::MissingPrivileges(errors))
+    } else {
+        Ok(())
+    }
+}
+
+/// Regex to parse a SHOW GRANTS line. Inspired by several stack overflow posts and
+/// adjusted to account for the different quoting styles across MySQL versions.
+/// If this regex matches then this is a grant on an actual object which looks like:
+///     GRANT SELECT, INSERT, UPDATE ON `db1`.* TO `u1`@`localhost`
+///     GRANT SELECT ON `db1`.`table1` TO `my_user`@`localhost` WITH GRANT OPTION
+/// The regex needs to account for the possibility of a wildcard schema or table, and for the
+/// quote-char to be part of the table/schema name too.
+/// Group 1 is the list of privileges being granted
+/// Group 2 is either the wildcard * or a quoted database/schema
+/// Group 4 is the unquoted database/schema when the wildcard is not matched
+/// Group 5 is either the wildcard * or a quoted table
+/// Group 7 is the unquoted table when the wildcard is not matched
+/// Group 9 is the user being granted
```

Comment:
You can name your groups to make the regular expression easier to read https://docs.rs/regex/latest/regex/#example-named-capture-groups

Reply:
Ah yeah I did write it with names at one point but thought the regex was even harder to understand that way (lots of additional `?<...>` and `\k<...>` statements)